homeassistant/components/vacuum/ecovacs.py

```diff
@@ -0,0 +1,222 @@
+"""
+Support for Ecovacs Ecovacs Vaccums.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/vacuum.neato/
+"""
+import logging
+
+from homeassistant.components.vacuum import (
+    VacuumDevice, SUPPORT_BATTERY, SUPPORT_RETURN_HOME, SUPPORT_CLEAN_SPOT,
+    SUPPORT_STATUS, SUPPORT_STOP, SUPPORT_TURN_OFF, SUPPORT_TURN_ON,
+    SUPPORT_LOCATE, SUPPORT_FAN_SPEED, SUPPORT_SEND_COMMAND, )
+from homeassistant.components.ecovacs import (
+    ECOVACS_DEVICES)
+from homeassistant.helpers.icon import icon_for_battery_level
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['ecovacs']
+
+SUPPORT_ECOVACS = (
+    SUPPORT_BATTERY | SUPPORT_RETURN_HOME | SUPPORT_CLEAN_SPOT |
+    SUPPORT_STOP | SUPPORT_TURN_OFF | SUPPORT_TURN_ON | SUPPORT_LOCATE |
+    SUPPORT_STATUS | SUPPORT_SEND_COMMAND | SUPPORT_FAN_SPEED)
+
+ECOVACS_FAN_SPEED_LIST = ['normal', 'high']
+
+# These consts represent bot statuses that can come from the `sucks` library
+STATUS_AUTO = 'auto'
+STATUS_EDGE = 'edge'
+STATUS_SPOT = 'spot'
+STATUS_SINGLE_ROOM = 'single_room'
+STATUS_STOP = 'stop'
+STATUS_RETURNING = 'returning'
+STATUS_CHARGING = 'charging'
+STATUS_IDLE = 'idle'
+STATUS_ERROR = 'error'
+
+# Any status that represents active cleaning
+STATUSES_CLEANING = [STATUS_AUTO, STATUS_EDGE, STATUS_SPOT, STATUS_SINGLE_ROOM]
+# Any status that represents sitting on the charger
+STATUSES_CHARGING = [STATUS_CHARGING, STATUS_IDLE]
+
+ATTR_ERROR = 'error'
+ATTR_COMPONENT_PREFIX = 'component_'
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Ecovacs vacuums."""
+    vacuums = []
+    for device in hass.data[ECOVACS_DEVICES]:
+        vacuums.append(EcovacsVacuum(device))
+    _LOGGER.debug("Adding Ecovacs Vacuums to Hass: %s", vacuums)
+    add_devices(vacuums, True)
+
+
+class EcovacsVacuum(VacuumDevice):
+    """Ecovacs Vacuums such as Deebot."""
+
+    def __init__(self, device):
+        """Initialize the Ecovacs Vacuum."""
+        self.device = device
+        self.device.connect_and_wait_until_ready()
+        try:
+            self._name = '{}'.format(self.device.vacuum['nick'])
+        except KeyError:
+            # In case there is no nickname defined, use the device id
+            self._name = '{}'.format(self.device.vacuum['did'])
+
+        self._fan_speed = None
+        self._error = None
+        _LOGGER.debug("Vacuum initialized: %s", self.name)
+
+    async def async_added_to_hass(self) -> None:
+        """d."""
+        # Fire off some queries to get initial state
+        self.device.statusEvents.subscribe(self.on_status)
```

Comment:
Since all you do is call `schedule_update_ha_state`, why not use lambdas?

```python
self.device.statusEvents.subscribe(lambda _: self.schedule_update_ha_state())
```

Reply:
Yep, that's better. Each of those event handlers used to do more, but I moved the logic to the `sucks` library since their creation. I'll make this change.